### ISSUE TITLE
Add/modify tool to swap Y and Z coordinates

### ISSUE THREAD

**Issue by CommunyAS** [qgis/QGIS#56309]
### Feature description

QGIS still is mainly 2D, so working (graphically) with geometries extending along the Z-axis has considerable limitations. For instance, in archaeology, to work with profile sections of features one has to rotate the profiles on the XY-plane (so that the geometries align along the X-axis) and then swap the Y and Z coordinates.

At the moment swapping the Y and Z coordinates is not possible in **QGIS** with the tools provided in the toolbox. Theoretically, it could be possible with the `GRASS v.transform` tool, but optional commands (like coordinate swapping) can only be applied from within the **GRASS** app, not from within **QGIS** (with the `GRASS v.transform` tool). Firing up the **GRASS** plugin in **QGIS** doesn’t help – I tried using the `GRASS shell` which, unfortunately, remains unresponsive so that one is not able to run `v.transform` code in the QGIS environment.
My workaround to swap Y and Z for the purpose described above is: 1) rotate and perhaps shift the geometries in QGIS with the `Affine transform` tool; 3) export the geometries (i.e. als GPKG; 2) Fire up the GRASS app, import the geometries and swap Y and Z within GRASS with optional commands of the `v.transform` tools; 3) export the swapped geometries from GRASS; 4) re-import them to QGIS.

I see three options to add swapping Y and Z with tools from the **QGIS** toolbox:

1. Extend the QGIS `Affine transform` tool - it basically seems identical to the GRASS `v.transform` tool, but is even better since one can use expressions/calculations native to QGIS to define the transformation.
2. Modify the `GRASS v.transform` tool in the QGIS toolbox so that optional commands become available (either by selecting them, like in **GRASS**, or be passing them as additional code).
3. Extend the QGIS `Swap X and Y coordinates` tool.

IMO option 1. would be the best, since it would provide users with the extensive functionality of QGIS (like using expressions, with is not possible with the `GRASS v.transform` tool).

(Perhaps there is also the option to fix the functionality of the `GRASS shell` of the **QGIS GRASS plugin** so that one could use the native GRASS `v.transform` tool with command line)

**Comment by agiudiceandrea** [comment]
@CommunyAS, thanks. May you please file separate bug report about the GRASS issues?

**Comment by agiudiceandrea** [comment]
@CommunyAS, indeed the "GRASS Shell" functionality has been removed https://github.com/qgis/QGIS/pull/53597. Anyway you also reported a bug in the v.transform GRASS algorithm which misses some flags of the v.transform GRASS module.

**Comment by CommunyAS** [comment]
@agiudiceandrea, I don't think this is a bug, but perhaps never was implemented in the GRASS `v.transform` tool within QGIS. Optional commands of the native GRASS `v.transform` were already not available at least about 5 years ago - and of you run GRASS within QGIS it is and has even been worse: `v.transform` then offers only a very limited set of 'regular' commands/options compared to what is available normally. Perhaps these screenshots demonstrate this better:

`v.transform` commands as QGIS GRASS tool

<img width="1103" alt="vtransform_QGIS_GRASStool" src="https://github.com/qgis/QGIS/assets/74900017/74e8684b-d2ca-4cc3-8b1d-38f95f8d6053">

`v.transform` commands of GRASS within QGIS

![vtransform_GRASS_QGIS](https://github.com/qgis/QGIS/assets/74900017/ceb35822-2ffd-4392-857c-880ec4b6ffc0)

`v.transform` 'regular' commands in native GRASS

![vtransform_GRASSnative_regular_commands](https://github.com/qgis/QGIS/assets/74900017/bef5edcd-da1a-4ad4-8bf6-b64e2e9df8d9)

`v.transform` optional commands in native GRASS

![vtransform_GRASSnative_optional](https://github.com/qgis/QGIS/assets/74900017/2c89ef02-390e-421b-92ca-89fc289bb738)

BTW, trying to work with native GRASS I realised that there are several problems with the `v.transform` and export tools that seem to be bugs. However, I (apart from the special issue of swapping Y and Z) exclusively work in QGIS.

**Comment by agiudiceandrea** [comment]
The v.transfom issue fixed with https://github.com/qgis/QGIS/pull/56337.

**Comment by agiudiceandrea** [comment]
The -t, -w, -x, -y, -a and -b missing parameters have been added to the v.transform processing algorithm by the PR https://github.com/qgis/QGIS/pull/56337. The fix will be available in QGIS >= 3.36, since this is a feature request.
Anyway you can manually fix the issue just replacing the v.transform.txt file in your QGIS installation folder with the patched one@ from that PR.

**Comment by CommunyAS** [comment]
This is great and proves my supposition wrong.

I replaced the v.transform.txt as suggested, all the optional/advanced parameters are now available.

However, running the `Swap coordinates y, z …` parameter on 3D layers (PolygonZ, PointZ) resulted in the following error:
> WARNING: Vector map <outputbdc2af71601342d18cdac5f55a4b54b1> is 3D. Use format specific layer creation options (parameter 'lco') to export <in 3D rather than 2D (default).

Trying with a PointZ layer, I stored the original Y value in a Yorg column and insert *zshift:Yorg* / *column="zshift:Yorg"* into the v.out.org output layer options (lco). This didn’t work, perhaps because I didn’t use the command to produce a 3D output correctly.

However, it would definitely not work this was with LineZ or PolygonZ geometries - i.e. it would always result in a swapped 2D output.

What must be done to swap Y and Z coordinates of 3D input and get a 3D output?

**Comment by agiudiceandrea** [comment]
You need to use the ESRI Shapefile format for the output layer.
